core/src/main/java/com/linecorp/armeria/internal/common/AbstractHttp2ConnectionHandler.java

```diff
@@ -172,7 +188,15 @@ public final void close(ChannelHandlerContext ctx, ChannelPromise promise) throw
             }
         }
 
-        super.close(ctx, promise);
+        // Send final GOAWAY frame with the latest stream ID.
+        goAway(ctx, connection().remote().lastStreamCreated()).addListener(future -> {
+            final Throwable cause = future.cause();
+            if (cause != null) {
+                logger.warn("{} final GOAWAY frame write completed exceptionally: {}", ctx.channel(), cause);
+            }
+            super.close(ctx, promise);
+        });
+        ctx.flush();
```

Comment:
Doesn't `super.close()` already send a GOAWAY frame with last known stream ID? See `io.netty.handler.codec.http2.Http2ConnectionHandler.close(ChannelHandlerContext, ChannelPromise)`.

Reply:
It does not if you already sent any GOAWAY for this connection. If we shutdown gracefully and send GOAWAY 2^31-1 `connection().goAwaySent()` will return true and `Http2ConnectionHandler.close` will `ctx.write(EMPTY_BUFFER)` instead of sending GOAWAY.

https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L475

Added it in `AbstractHttp2ConnectionHandler` instead of `Http2ServerConnectionHandler` to make sure that we don't need to worry about those implementation details in the future, e.g. if we add support for graceful shutdown to HTTP/2 client in the future.

Reply:
SGTM. Thanks for elaboration.